plugins/github/api.ml

```diff
@@ -94,36 +94,68 @@ module Status = struct
   type t = {
       state: state;
-      description: string option;
+      summary: string option;
+      text: string option; 
+      actions: action list;
       url: Uri.t option
     }
 
-  let v ?description ?url state =
-    let description = Option.map (Astring.String.with_range ~len:140) description in (* Max GitHub allows *)
-    { state; description; url }
+  let action ~label ~description ~identifier = {label; description; identifier}
 
-  let state_to_string = function
+  let action_to_json {label; description; identifier} =
+    `Assoc [ ("label", `String label)
+           ; ("description", `String description)
+           ; ("identifier", `String identifier)]
+
+  let v ?text ?summary ?url ?(actions=[]) state =
+    (* A maximum of three actions are accepted by GitHub. *)
+    { state; summary; text; actions; url }
+
+  let state_json ~title ~summary ~text ~status () =
+    let output = [ ("title", `String title)
+                 ; ("summary", `String summary) ] @
+                 (match text with None -> [] | Some x -> ["text", `String x])
+    in
+    [ "status", `String status
+    ; "output", `Assoc output ]
+
+  let json_items { state; text; summary; actions; url } =
+    (match url with None -> [] | Some url -> ["details_url", `String (Uri.to_string url)]) @
+    match state with
     | `Queued -> 
-       ["status", `String "queued"]
+      state_json ~title:"Queued"
+                 ~summary:(Option.value summary ~default:"Queued")
+                 ~status:"queued" ~text ()
+
     | `InProgress -> 
-       ["status", `String "inprogress"]
-    | `Completed `Success -> 
-       ["status", `String "completed"
-       ;"conclusion", `String "success"]
-    | `Completed (`Failure msg) -> 
-       ["status", `String "completed"
-       ;"conclusion", `String "failure"
-       ;"output", `Assoc [("title", `String "Failure")
-                         ;("summary", `String msg)]]
+      state_json ~title:"InProgress" 
+                 ~summary:(Option.value summary ~default:"InProgress")
+                 ~status:"inprogress" ~text ()
 
-  let json_items { state; description; url } =
-    state_to_string state @
-    (match description with None -> [] | Some x -> ["description", `String x]) @
-    (match url with None -> [] | Some x -> ["details_url", `String (Uri.to_string x)])
+    | `Completed `Success -> 
+      ("conclusion", `String "success") ::
+      state_json ~title:"Completed" 
+                 ~summary:(Option.value summary ~default:"Completed")
+                 ~status:"completed" ~text ()
+
+    | `Completed (`Failure _) ->
+       [ "conclusion", `String "failure"
+       ; "actions", `List (List.map action_to_json actions) ] @ 
+       state_json ~title:"Failure" 
+                  ~summary:(Option.value summary ~default:"Failure")
+                  ~status:"completed" ~text ()
+
+    | `Completed (`Skipped _) -> 
+      [ "conclusion", `String "skipped"
```

Comment:
BTW, the idea with the `state_json` was to keep all the magic strings in one place. I just didn't read far enough to see that there were more. Perhaps it should have been:

```ocaml
  let state_json ~title ~summary ~text ~status ?conclusion ?(actions=[]) () =
    let output = [ ("title", `String title)
                 ; ("summary", `String summary) ] @
                 (match text with None -> [] | Some x -> ["text", `String x])
    in
    [ "status", `String status
    ; "output", `Assoc output ] @
    (if actions = [] then [] else ["actions", `List (List.map action_to_json actions)]) @
    (match conclusion with None -> [] | Some x -> ["conclusion", `String x])
```